src/tbox/stream/impl/stream/buffer.c

```diff
@@ -0,0 +1,254 @@
+ */
+#include "prefix.h"
+
+/* //////////////////////////////////////////////////////////////////////////////////////
+ * types
+ */
+
+// the buffer stream type
+typedef struct __tb_stream_buffer_t
+{
+    // the data 
+    tb_buffer_ref_t         buffer;
+
+    // the head
+    tb_size_t               head;
+
+    // the data is referenced?
+    tb_bool_t               bref;
+
+}tb_stream_buffer_t;
+
+/* //////////////////////////////////////////////////////////////////////////////////////
+ * implementation
+ */
+static __tb_inline__ tb_stream_buffer_t* tb_stream_buffer_cast(tb_stream_ref_t stream)
+{
+    // check
+    tb_assert_and_check_return_val(stream && tb_stream_type(stream) == TB_STREAM_TYPE_BUFF, tb_null);
+
+    // ok?
+    return (tb_stream_buffer_t*)stream;
+}
+static tb_bool_t tb_stream_buffer_open(tb_stream_ref_t stream)
+{
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return_val(stream_buffer, tb_false);
+    
+    stream_buffer->head = 0;
+
+    // ok
+    return tb_true;
+}
+static tb_bool_t tb_stream_buffer_clos(tb_stream_ref_t stream)
+{
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return_val(stream_buffer, tb_false);
+
+    // clear head
+    stream_buffer->head = 0;
+
+    // ok
+    return tb_true;
+}
+static tb_void_t tb_stream_buffer_exit(tb_stream_ref_t stream)
+{
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return(stream_buffer);
+
+    // clear head
+    stream_buffer->head = 0;
+
+    // exit data
+    if (stream_buffer->buffer && !stream_buffer->bref) tb_buffer_exit(stream_buffer->buffer);
+    stream_buffer->buffer = tb_null;
+}
+static tb_long_t tb_stream_buffer_read(tb_stream_ref_t stream, tb_byte_t* data, tb_size_t size)
+{
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return_val(stream_buffer && stream_buffer->buffer, -1);
+
+    // check
+    tb_check_return_val(data, -1);
+    tb_check_return_val(size, 0);
+
+    // the left
+    tb_size_t left = tb_buffer_size(stream_buffer->buffer) - stream_buffer->head;
+
+    // the need
+    if (size > left) size = left;
+
+    // read data
+    if (size) tb_memcpy(data, tb_buffer_data(stream_buffer->buffer) + stream_buffer->head, size);
+
+    // save head
+    stream_buffer->head += size;
+
+    // ok?
+    return (tb_long_t)(size);
+}
+static tb_long_t tb_stream_buffer_writ(tb_stream_ref_t stream, tb_byte_t const* data, tb_size_t size)
+{
+    // check
+    tb_stream_buffer_t* stream_buffer = tb_stream_buffer_cast(stream);
+    tb_assert_and_check_return_val(stream_buffer && stream_buffer->buffer, -1);
+
+    // check
+    tb_check_return_val(data, -1);
+    tb_check_return_val(size, 0);
+
+    // writ data
+    if (size) tb_buffer_memncpyp(stream_buffer->buffer, stream_buffer->head, data, size);
```

Comment:
I noticed that the xml writer write one byte to do the formatting. Is the buffer efficient in being reallocated regularly? If not maybe we could resize the buffer to contains more than what is it written to it, a bit like a vector in C++

Reply:
not need, stream has write cache. https://github.com/tboox/tbox/blob/3fdd8f0ab91cea9fe516ab31035ea79425c04c4a/src/tbox/stream/stream.c#L917